### ISSUE TITLE
Define `@trymatch` and `@tryswitch`

### ISSUE THREAD

**Issue by serenity4** [thautwarm/MLStyle.jl#131]
I think it would be a good addition to have versions of `@trymatch` and `@tryswitch` that allow having no matches, like for example

```julia

@trymatch x begin
  ::Int => :int
end

# would be equivalent to

@match x begin
  ::Int => :int
  _ => nothing
end
```

and

```julia
@tryswitch x begin
  @case ::Int
    println("int")
end

# would be equivalent to

@switch x begin
  @case ::Int
    println("int")
  @case _
    nothing
end
```

The main advantage with these is when you have nested `@match` and `@switch` statements, where each invocation will require adding this special case for no matches. The name should make it fairly obvious that it will silently "fail".

If it's not already implemented (I looked around, but we never know) I can submit a PR to define these helper macros.

**Comment by thautwarm** [comment]
Hello, what you said is really a practical feature and we support it via a more concise syntax `@when` and `@otherwise`:

```julia
julia> @when (a, b, 1) = (1, 2, 1) begin # the left-hand side is a pattern, and the right is a value
           println(a, b)
       end

12

julia> @when (a, b, 1) = (1, 2, 2) begin
           println(a, b)
       end
```

You can also have more `@when` clauses and an optional `@otherwise` clause:

```julia
julia> test = (1, 2, 3)

(1, 2, 3)

julia> @when (a, b) = test begin
           println(1)
       @when (a, b, c, d) = test
           println(2)
       @otherwise
           println(3)
       end
3

```

**Comment by serenity4** [comment]
It seems very nice when you want to specify successive matches, but doesn't cover the `@match`/`@switch` logic where every clause refers to a specific match.

Consider this example with `@match` and with `@when`:
```julia
@match x begin
  ::String => :string
  ::Int => :int
  ::Bool => :bool
  ::Nothing => :nothing
  _ => nothing
end

@when ::String = test begin;
    :string
    @when ::Int = test; :int
    @when ::Bool = test; :bool
    @when ::Nothing = test; :nothing
end
```
There is a redundancy in having to always specify `= test`, that could be avoided.

You can take a look at a concrete use case [here](https://github.com/serenity4/SPIRV.jl/blob/master/src/ir.jl#L87-L194) where it came in really handy.

**Comment by thautwarm** [comment]
Thanks for this concrete use case! It looks fine to me, and I can add `@trymatch` later.

Or maybe @Roger-luo could add such functionality to a 3rd-party utility lib?

**Comment by serenity4** [comment]
I think we can reopen this, since `@trymatch` still hasn't been implemented yet. But I can take a shot at it, it should be similar to #132.

**Comment by thautwarm** [comment]
Okay, just reopen it.

**Comment by serenity4** [comment]
Implemented in #133  and #132.
